[GPU Process] Allow replaying back a whole DisplayList in GPU Process

shallawa opened the pull request:
#### c708ae60f9fa9abf1d383d35fccb149786f50a54
<pre>
[GPU Process] Allow replaying back a whole DisplayList in GPU Process
<a href="https://bugs.webkit.org/show_bug.cgi?id=260181">https://bugs.webkit.org/show_bug.cgi?id=260181</a>
<a href="https://rdar.apple.com/113878782">rdar://113878782</a>

Reviewed by Cameron McCormack.

Add a new DisplayList item called DrawDisplayListItems. Add the needed functions
for recording it and replaying it back. The ResourceHeap will not be stored in
this item. Instead it will be passed as an extra argument to its apply() method.

Add the new method GraphicsContext::drawDisplayListItems() which draws a DisplayList
given ResourceHeap and a destination point.

In Recorder::drawDisplayListItems() record the use all the ResourceHeap resources
and then call recordDrawDisplayListItems().

In RecorderImpl::recordDrawDisplayListItems(), add the item DrawDisplayListItems
to the recording DisplayList.

In RemoteDisplayListRecorderProxy::recordDrawDisplayListItems(), send the message
RemoteDisplayListRecorder.DrawDisplayListItem to GPUProcess.

In RemoteDisplayListRecorder::drawDisplayListItems(), create a DrawDisplayListItems
item given the received DisplayList. Then call its apply() method given the
ResourceHeap of the RemoteResourceCache.

TextPainter will call GraphicsContext::drawDisplayListItems() instead of replaying
the cached DisplayList back.

All DisplayList items have to have ArgumentCoders because DrawDisplayListItems
includes a Vector of DisplayListItem which is a std::variant of all the items.

* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawDisplayListItems):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawDisplayListItems::DrawDisplayListItems):
(WebCore::DisplayList::DrawDisplayListItems::apply const):
(WebCore::DisplayList::DrawDisplayListItems::dump const):
(WebCore::DisplayList::PaintFrameForMedia::PaintFrameForMedia):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawFilteredImageBuffer::filter const):
(WebCore::DisplayList::DrawDisplayListItems::items const):
(WebCore::DisplayList::DrawDisplayListItems::destination const):
(WebCore::DisplayList::DrawNativeImage::imageSize const):
(WebCore::DisplayList::DrawNativeImage::source const):
(WebCore::DisplayList::DrawNativeImage::options const):
(WebCore::DisplayList::DrawPattern::options const):
(WebCore::DisplayList::FillLine::line const):
(WebCore::DisplayList::FillArc::arc const):
(WebCore::DisplayList::FillQuadCurve::quadCurve const):
(WebCore::DisplayList::FillBezierCurve::bezierCurve const):
(WebCore::DisplayList::FillPathSegment::segment const):
(WebCore::DisplayList::PaintFrameForMedia::destination const):
(WebCore::DisplayList::StrokeArc::arc const):
(WebCore::DisplayList::StrokeQuadCurve::quadCurve const):
(WebCore::DisplayList::StrokeBezierCurve::bezierCurve const):
(WebCore::DisplayList::StrokePathSegment::segment const):
(WebCore::DisplayList::DrawControlPart::part const):
(WebCore::DisplayList::DrawControlPart::type const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::drawDisplayListItems):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordDrawDisplayListItems):
(WebCore::DisplayList::RecorderImpl::recordPaintFrameForMedia):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListReplayer.h:
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::TextPainter::paintTextOrEmphasisMarks):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawDisplayListItems):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
(WebKit::RemoteResourceCache::resourceHeap const):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawDisplayListItems):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/270688@main">https://commits.webkit.org/270688@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69bc3ea8195e63c562d7a9b95d2bc6757c6ffaf0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26081 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27355 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28177 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23882 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26404 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6456 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2109 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23934 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3553 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28753 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3168 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23411 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29464 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23786 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23792 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27348 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1406 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4605 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6283 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3669 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3528 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -236,6 +236,35 @@ void Recorder::drawGlyphsAndCacheResources(const Font& font, const GlyphBufferGl
+
+        if (auto font = std::get_if<Ref<Font>>(&resource.value))
+            recordResourceUse(font->get());
+    }
```
Can we use a `WTF::switchOn` for `resource`? That way we'll get a compile error if we add new types to the `ResourceHeap::Resource` variant. That won't help with the `RenderingResource` types — maybe we can either add an assertion in here that we don't end up not handling the resource, or we could add a virtual function to `RenderingResource` to add it to a given `DisplayList::Recorder`?

heycam commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -236,6 +236,35 @@ void Recorder::drawGlyphsAndCacheResources(const Font& font, const GlyphBufferGl
+            continue;
+        }
+
+        if (auto font = std::get_if<Ref<Font>>(&resource.value))
```
Nit: `auto* font`, if you want to be consistent with the previous `if` statements.

heycam commented on Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
```diff
@@ -20,17 +20,90 @@
+    WebCore::FloatSize amount();
+};
+
+[CustomHeader, AdditionalEncoder=StreamConnectionEncoder] class WebCore::DisplayList::SetCTM {
+    WebCore::AffineTransform transform();
+};
```
This reminds me: if in a future patch we support recording arbitrary display lists, then we'll need to do something like `DrawGlyphsRecorder::updateCTM`, where we effectively convert `setCTM()` calls into equivalent `concatCTM()`s, so that the display list can be replayed at different translations.

heycam commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
```diff
@@ -626,6 +653,8 @@ class DrawPattern {
     const AffineTransform& patternTransform() const { return m_patternTransform; }
     FloatPoint phase() const { return m_phase; }
     FloatSize spacing() const { return m_spacing; }
+    const ImagePaintingOptions& options() const { return m_options; }
```
`ImagePaintingOptions` seems to be designed to pass by value, so we probably don't need to return a const reference here.

heycam commented on Source/WebCore/platform/graphics/GraphicsContext.h:
```diff
@@ -314,6 +316,9 @@ class GraphicsContext {
     virtual void drawDotsForDocumentMarker(const FloatRect&, DocumentMarkerLineStyle) = 0;
 
+    // DisplayList
+    WEBCORE_EXPORT virtual void drawDisplayListItems(const Vector<DisplayList::Item>&, const DisplayList::ResourceHeap&, const FloatPoint& destination);
```
It feels a bit strange to pass in a `ResourceHeap` to a `GraphicsContext` drawing function, like it's almost an implementation detail that the `GraphicsContext` layer shouldn't need to know about. But I don't have a better suggestion right now, since changing this to take a `DisplayList` instead would mean that when replaying in the GPU process, we'd need to construct a `DisplayList` from the recorded items plus a reference to the destination's `ResourceHeap`. But `DisplayList` only supports holding an owned copy of the resource heap.

Maybe in a future patch if you move towards display lists being a resource type themselves, we could have the reusable display list own and transfer its resource heap, instead of recording resource use in the target display list.

heycam commented on Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
```diff
@@ -20,17 +20,90 @@
+headers: <WebCore/DisplayListItem.h>
 headers: <WebCore/DisplayListItems.h>
 
+[CustomHeader, AdditionalEncoder=StreamConnectionEncoder] class WebCore::DisplayList::Save {
```
I did not closely review the serialization changes here.

shallawa committed:
[GPU Process] Allow replaying back a whole DisplayList in GPU Process
https://bugs.webkit.org/show_bug.cgi?id=260181
rdar://113878782

Reviewed by Cameron McCormack.

Add a new DisplayList item called DrawDisplayListItems. Add the needed functions
for recording it and replaying it back. The ResourceHeap will not be stored in
this item. Instead it will be passed as an extra argument to its apply() method.

Add the new method GraphicsContext::drawDisplayListItems() which draws a DisplayList
given ResourceHeap and a destination point.

In Recorder::drawDisplayListItems() record the use all the ResourceHeap resources
and then call recordDrawDisplayListItems().

In RecorderImpl::recordDrawDisplayListItems(), add the item DrawDisplayListItems
to the recording DisplayList.

In RemoteDisplayListRecorderProxy::recordDrawDisplayListItems(), send the message
RemoteDisplayListRecorder.DrawDisplayListItem to GPUProcess.

In RemoteDisplayListRecorder::drawDisplayListItems(), create a DrawDisplayListItems
item given the received DisplayList. Then call its apply() method given the
ResourceHeap of the RemoteResourceCache.

TextPainter will call GraphicsContext::drawDisplayListItems() instead of replaying
the cached DisplayList back.

All DisplayList items have to have ArgumentCoders because DrawDisplayListItems
includes a Vector of DisplayListItem which is a std::variant of all the items.

* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::drawDisplayListItems):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyItem):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawDisplayListItems::DrawDisplayListItems):
(WebCore::DisplayList::DrawDisplayListItems::apply const):
(WebCore::DisplayList::DrawDisplayListItems::dump const):
(WebCore::DisplayList::PaintFrameForMedia::PaintFrameForMedia):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::DrawFilteredImageBuffer::filter const):
(WebCore::DisplayList::DrawDisplayListItems::items const):
(WebCore::DisplayList::DrawDisplayListItems::destination const):
(WebCore::DisplayList::DrawNativeImage::imageSize const):
(WebCore::DisplayList::DrawNativeImage::source const):
(WebCore::DisplayList::DrawNativeImage::options const):
(WebCore::DisplayList::DrawPattern::options const):
(WebCore::DisplayList::FillLine::line const):
(WebCore::DisplayList::FillArc::arc const):
(WebCore::DisplayList::FillQuadCurve::quadCurve const):
(WebCore::DisplayList::FillBezierCurve::bezierCurve const):
(WebCore::DisplayList::FillPathSegment::segment const):
(WebCore::DisplayList::PaintFrameForMedia::destination const):
(WebCore::DisplayList::StrokeArc::arc const):
(WebCore::DisplayList::StrokeQuadCurve::quadCurve const):
(WebCore::DisplayList::StrokeBezierCurve::bezierCurve const):
(WebCore::DisplayList::StrokePathSegment::segment const):
(WebCore::DisplayList::DrawControlPart::part const):
(WebCore::DisplayList::DrawControlPart::type const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::drawDisplayListItems):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordDrawDisplayListItems):
(WebCore::DisplayList::RecorderImpl::recordPaintFrameForMedia):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListReplayer.h:
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::TextPainter::paintTextOrEmphasisMarks):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawDisplayListItems):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
(WebKit::RemoteResourceCache::resourceHeap const):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordDrawDisplayListItems):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: https://commits.webkit.org/270688@main
